Stark Ramsey experiment

nkanazawa1989 opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add it in the CHANGELOG file under Unreleased section.
⚠️ If your pull request fixes an open issue, please link to the issue.

✅ I have added the tests to cover my changes.
✅ I have updated the documentation accordingly.
✅ I have read the CONTRIBUTING document.
-->

### Summary

This PR adds new experiment class `StarkRamseyXY`. Usually fixed-qubit frequency architecture such as one of IBM Quantum backend doesn't provide a control knob of qubit frequency. However, we can still scan qubit frequency by applying an off-resonant drive that causes the Stark shift on the qubit.

New experiment characterizes the amount of Stark shift at certain drive amplitude. 

### Details and comments

The test experiment result is available in [546871be-e200-469b-bba7-ae49aefc3c78](https://quantum-computing.ibm.com/experiments/546871be-e200-469b-bba7-ae49aefc3c78). This experiment result was obtained with the following experiment code:

```python
from qiskit_experiments.library import StarkRamseyXY

channel = backend.configuration().control((0, 1))[0]
exp = StarkRamseyXY(0, 0.1, backend)
exp.set_experiment_options(stark_channel=channel)

exp_data = exp.run().block_for_results()
```

Note that IBM backends update the channel frame with set_frequency or shift_frequency command. Thus one of control channel is assigned to the Stark drive channel not to mixup the qubit frame.

wshanks commented on qiskit_experiments/library/stark/ramsey_xy.py:
```diff
@@ -0,0 +1,319 @@
+
+            .. _stark_amplitude:
+
+            The `stark_channel` in the experiment options may be necessary supplied.
+            By default, the Stark tone is applied to the same channel with the qubit drive
+            with frequency modulation. This frequency modulation might update the
+            channel frame, which accumulates unwanted phase against the frequency difference
+            between the qubit drive and Stark tone frequency in addition to
+            the qubit Stark shift. You can use a dedicated Stark drive channel if available,
+            otherwise you may want to reuse a control channel associated with the qubit.
+            This depends on the architecture of your quantum device.
```
```suggestion
            It may be necessary to supply the ``stark_channel`` experiment option.
            By default, the Stark tone is applied to the same channel as the qubit drive
            with a frequency shift. This frequency shift might update the
            channel frame, which accumulates unwanted phase against the frequency difference
            between the qubit drive and Stark tone frequencies in addition to
            the qubit Stark shift. You can use a dedicated Stark drive channel if available.
            Otherwise you may want to use a control channel associated with the physical
            drive port of the qubit (e.g. in a cross-resonance system, a control channel for
            which the qubit is the control qubit).
            This depends on the architecture of your quantum device.
```

wshanks commented on qiskit_experiments/library/stark/ramsey_xy.py:
```diff
@@ -0,0 +1,319 @@
+
+        Args:
+            qubit: Index of qubit.
+            stark_amp: A single float parameter to represent the magnitude of Stark tone
+                and the sign of expected Stark shift. See :ref:`stark_amplitude` for details.
```
I think this should reference the class-level note instead.

```suggestion
            stark_amp: A single float parameter to represent the magnitude of the Stark tone
                and the sign of the expected Stark shift. See :ref:`stark_amplitude` for details.
```

wshanks commented on qiskit_experiments/library/stark/ramsey_xy.py:
```diff
@@ -0,0 +1,319 @@
+            stark_amp: A single float parameter to represent the magnitude of Stark tone
+                and the sign of expected Stark shift. See :ref:`stark_amplitude` for details.
+            backend: Optional, the backend to run the experiment on.
+            experiment_options: Extra experiment options. See self.experiment_options.
```
Is there a better way to reference the experiment options? In code, you would look at `_default_experiment_options`. In Sphinx, they have a section of the experiment documentation.

```suggestion
            experiment_options: Extra experiment options. See ``self.experiment_options``.
```

nkanazawa1989 replied:
The documentation there is automatically copied to the class doc with the experiment sphinx macro, however I didn't insert any reference pointer to there. So currently we cannot directly refer to the document.

wshanks commented on qiskit_experiments/library/stark/ramsey_xy.py:
```diff
@@ -0,0 +1,319 @@
+                Note that fitter algorithm :class:`.RamseyXYAnalysis` of this experiment
+                is still capable of fit experiment data with further lower frequency.
+            max_freq (float): Maximum frequency that this experiment can resolve.
+            delays (list[float]): The list of delays in set that will be scanned in the
```
```suggestion
            delays (list[float]): The list of delays if set that will be scanned in the
```

wshanks commented on qiskit_experiments/library/stark/ramsey_xy.py:
```diff
@@ -0,0 +1,319 @@
+            delays (list[float]): The list of delays in set that will be scanned in the
+                experiment. If not set, then evenly spaced delays with interval
+                computed from ``min_freq`` and ``max_freq`` are used.
+                If ``delays`` is set, these parameters are ignored.
```
This seems to say the same thing as the first sentence.

```suggestion
```

wshanks commented on qiskit_experiments/library/stark/ramsey_xy.py:
```diff
@@ -0,0 +1,319 @@
+                is still capable of fit experiment data with further lower frequency.
+            max_freq (float): Maximum frequency that this experiment can resolve.
+            delays (list[float]): The list of delays in set that will be scanned in the
+                experiment. If not set, then evenly spaced delays with interval
```
Should there be a reference to the `delays` method which gives more description?

wshanks commented on qiskit_experiments/library/stark/ramsey_xy.py:
```diff
@@ -0,0 +1,319 @@
+
+        .. note::
+
+            The delays are computed with min_freq and max_freq experiment options.
+            The maximum point is computed from the min_freq to guarantee the result
+            contains at least one Ramsey oscillation cycle at this frequency.
+            The interval is computed from the max_freq to sample at frequency
+            higher than the Nyquist frequency of this frequency.
```
```suggestion
            The delays are computed with the ``min_freq`` and ``max_freq`` experiment options.
            The maximum point is computed from the ``min_freq`` to guarantee the result
            contains at least one Ramsey oscillation cycle at this frequency.
            The interval is computed from the ``max_freq`` to sample with resolution
            such that the Nyquist frequency is twice ``max_freq``.
```

wshanks commented on releasenotes/notes/mod-stark-1f1afb538a94fe9a.yaml:
```diff
@@ -0,0 +1,12 @@
+  - |
+    New module :mod:`qiskit_experiments.library.stark` has been added. 
+    This module contains characterization experiments utilizing the Stark shift
+    for spectroscopy. This allows experimentalist to characterize qubit parameter
```
```suggestion
    for spectroscopy. This allows an experimentalist to characterize qubit parameter
```

wshanks commented on qiskit_experiments/library/stark/ramsey_xy.py:
```diff
@@ -0,0 +1,319 @@
+    def __init__(
+        self,
+        qubit: int,
+        stark_amp: float,
```
I wonder if `stark_amp` is really more special than the other options and needs to be split out as a required parameter?

nkanazawa1989 replied:
Yes, I think so. This experiment intends to measure RamseyXY at a particular qubit frequency that is determined by the stark_amp. If we provide random default value, user will never know what they are measuring.

wshanks commented on test/library/stark/test_ramsey_xy.py:
```diff
@@ -0,0 +1,125 @@
+            )
+
+        test_schedule = param_ram_x.calibrations["StarkV"][(0,), ()]
+        self.assertEqual(test_schedule, ref_schedule)
```
This is fine, but I wonder if it would be cleaner to check that `test_schedule` starts with the right `set_frequency` instruction instead of comparing the full schedules?

nkanazawa1989 replied:
I also wanted to check if amp if always-positive so I needed to check full schedule. I also added the comment to tests.

wshanks replied:
Ah, okay, though a negative amp shouldn't hurt anything :slightly_smiling_face:

nkanazawa1989 replied:
Yes, exactly. At the same time I also found a cubit term was sometimes introduced at high power, so this prevents user from using unexpected frequency. And this is a strict form the amp defined in the manual.

wshanks replied:
Hmm, what prevents the user from using an unexpected frequency? I didn't follow the connection between the amp sign and the cubic term.

wshanks commented on test/library/stark/test_ramsey_xy.py:
```diff
@@ -0,0 +1,125 @@
+
+    def test_gen_delays(self):
+        """Test generating delays with experiment options."""
+        exp = StarkRamseyXY(
+            qubit=0,
+            stark_amp=0.1,
+            min_freq=1e6,
+            max_freq=50e6,
+        )
+        test_delays = exp.delays()
+        ref_delays = np.arange(0, 1e-6, 1e-8)
```
This might help someone looking at this later:

```suggestion
        min_freq = 1e6
        max_freq = 50e6
        exp = StarkRamseyXY(
            qubit=0,
            stark_amp=0.1,
            min_freq=min_freq,
            max_freq=max_freq,
        )
        test_delays = exp.delays()
        ref_delays = np.arange(0, 1 / min_freq, 1 / max_freq / 2)
```

wshanks reviewed (changes requested):
Looks good! Sorry, I put a lot of wording suggestions in.

nkanazawa1989 commented on qiskit_experiments/library/stark/ramsey_xy.py:
```diff
@@ -0,0 +1,319 @@
+                If ``delays`` is set, these parameters are ignored.
+        """
+        options = super()._default_experiment_options()
+        options.update_options(
```
`stark_freq_offset` doesn't control the circuit number. This is maybe confusing wording choice if you know about implementation of the Ramsey experiment. In principle you cannot place Stark tone at around qubit frequency or around f02/2 transition to avoid frequency collision. 80 MHz is typical offset value in this experiment, but user should know the mechanism behind the scene.

wshanks replied:
I was making two separate suggestions.

The first suggestion was warning about `stark_freq_offset < 1 / stark_sigma` because if it is too close to the qubit then the ramps could excite the qubit.

The second suggestion was about `min_freq` and `max_freq` . The number of circuits is `2 max_freq / min_freq` which I thought might be tricky for the user to reason about. I was wondering about warning if there are many circuits (like 600?).

nkanazawa1989 replied:
Good suggestions. Done in [7745034](https://github.com/Qiskit/qiskit-experiments/pull/1008/commits/774503448fd9bd25363638128ccd5cbe94a214c2) ... you may want to revise the message :P

nkanazawa1989 commented on qiskit_experiments/library/stark/ramsey_xy.py:
```diff
@@ -0,0 +1,319 @@
+
+        .. parsed-literal::
+
+            (Ramsey X)  The pulse before measurement rotates by pi-half around the X axis
```
Yes, usually ramps are problematic to do better fit. This is not super relevant to this experiment, but in the following PR I introduce `StarkRamseyFast` where we need to fit frequency shift against drive amplitude. Note that the fit function is already complicated, which is basically in the form of sin(x**2) and consideration of offset add more unwanted degree of freedom to the fit. This technique works really nicely with this experiment in my experience. Probably we can use conventional single tone for this experiment.

I haven't tried balancing the delay, but as you mentioned, the T2* dephasing is not a serious issue and even if it suffers the dephasing the fitter can still decouple that effect from frequency of interest by exponential decay envelope.

nkanazawa1989 replied:
Since a user may want to try T2* spectroscopy, probably it's better to use single tone here.

nkanazawa1989 commented on qiskit_experiments/library/stark/__init__.py:
```diff
@@ -0,0 +1,30 @@
+
+"""
+===============================================================================================
+Stark Experiments (:mod:`qiskit_experiments.library.stark`)
```
The purpose of this module is to provide experiment to deal with TLS. I don't think we will have many experiments in this module. So far I have implemented RamseyXY, RamseyXY with amplitude scan at fixed delay (StarkRamseyFast), and P1 spectroscopy. Probably we can put these experiments in the standard characterization module.

nkanazawa1989 replied:
I removed the dedicated module with [56349b8](https://github.com/Qiskit/qiskit-experiments/pull/1008/commits/56349b8ed6d7e9520dca17fbd0d0da7a4052893c).

wshanks replied:
Okay, that will probably be fine. I was not opposed to a new directory if there were enough experiments for it. I was wondering if we could scope it to a topic of study (like TLS, or maybe frequency shifting?) rather than a technique like Stark.

nkanazawa1989 commented on qiskit_experiments/library/stark/ramsey_xy.py:
```diff
@@ -0,0 +1,319 @@
+
+        Experiment Options:
+            stark_amp (float): A single float parameter to represent the magnitude of Stark tone
+                and the sign of expected Stark shift. See :ref:`stark_amplitude` for details.
```
I created a dedicated tutorial in [afd62a8](https://github.com/Qiskit/qiskit-experiments/pull/1008/commits/afd62a804490a4fd5e6b4527d732a755ce4a6ee9) because this is likely referred by every Stark experiments.

nkanazawa1989 commented on qiskit_experiments/library/stark/ramsey_xy.py:
```diff
@@ -0,0 +1,319 @@
+        stark_amp = np.abs(opt.stark_amp)
+        stark_channel = opt.stark_channel or pulse.DriveChannel(self.physical_qubits[0])
+        ramps_dt = self._timing.round_pulse(time=2 * opt.stark_risefall * opt.stark_sigma)
+        sigma_dt = ramps_dt / 2 / opt.stark_risefall
```
Rounding risefall is only option. Note that StarkV and U must have the identical edge shape to cancel out the time-dependent Stark shift. StarkV has the duration of the risefall.

wshanks replied:
StarkV has the duration of `round_pulse(2 * sigma * risefall)`. We can think of `2 * sigma * risefall` as the ideal width of StarkV. When it gets rounded, we can either stretch sigma by the same factor to keep `risefall` constant or we can keep `sigma` fix and let the stretched width redefine `risefall`. So either we redefine `sigma` to `sigma_dt` like you do or we redefine `risefall`. To do that, we would keep `sigma` fixed for both pulses but pass `ramps_dt / 2 / sigma` as `risefall_sigma_ratio` to StarkU. Keeping sigma fixed seems slightly better to me, but it probably is such a small difference that either is fine.

wshanks replied:
Here is the code I was using to compare the cases.

```python
from matplotlib import pyplot as plt
from qiskit import pulse


amp = 1
sigma = 13
width = 30
risefall = 1.25

g_duration = 10 * round(risefall * sigma * 2 / 10)

risefall = g_duration / 2 / sigma
#sigma = g_duration / risefall / 2
print(risefall)
print(sigma)

g = pulse.Gaussian(g_duration, amp, sigma)
gs = pulse.GaussianSquare(g_duration + width, amp, sigma, risefall_sigma_ratio=risefall)

fig = plt.figure()
ax = fig.add_subplot()
ax.plot(g.get_waveform().samples, "o-")
ax.plot(gs.get_waveform().samples, "o-")
ax.set_xlim(0, g_duration / 2)
```

nkanazawa1989 replied:
Thanks. I decided to update `BackendTiming` which looks simpler. I was more concerning the pulse renormalization (lifted Gaussian) at small sigma since it modifies spectrum of the pulse, but these are usually tiny difference as you mention :)

nkanazawa1989 commented on qiskit_experiments/library/stark/ramsey_xy.py:
```diff
@@ -0,0 +1,319 @@
+        for delay in self.delays():
+            # Not using pulse_round method of the BackendTiming class
+            # because this method considers the minimum pulse duration.
+            # Valid delay here corresponds to the flat-top length and thus can be zero at minimum.
```
Yes, this is another option, but more strictly this is `pulse_round(ramps_dt * dt + delay) - ramps_dt` which is bit complicated. Alternatively we can add a flag to pulse_round that controls minimum delay.

wshanks replied:
I don't feel strongly. Adding an option to `round_pulse` to ignore the minimum delay seems okay to me. I think it is nice to use `round_pulse` when we can since it allows the details of the timing constraints to change without affecting the experiment code.

nkanazawa1989 replied:
Fair enough. Done in [67ca32d](https://github.com/Qiskit/qiskit-experiments/pull/1008/commits/67ca32de0f1a69f80cac233ddee01d41f391b6a2). Instead of adding constraint argument to every method, I added them to the constructor so that we can avoid writing tons of conditional branch inside these methods.

nkanazawa1989 commented on qiskit_experiments/library/stark/ramsey_xy.py:
```diff
@@ -0,0 +1,319 @@
+        This technique allows the experiment to only accumulate the net phase
+        during the flat-top part of the StarkU pulse with the constant magnitude.
+
+    # section: note
```
I created dedicated tutorial for this in [afd62a8](https://github.com/Qiskit/qiskit-experiments/pull/1008/commits/afd62a804490a4fd5e6b4527d732a755ce4a6ee9).

nkanazawa1989 commented on qiskit_experiments/library/stark/__init__.py:
```diff
@@ -0,0 +1,30 @@
+    :toctree: ../stubs/
+    :template: autosummary/experiment.rst
+
+    StarkRamseyXY
```
I think this is the issue of docs tree design. Probably we can reorganize this in the followup.

nkanazawa1989 commented on qiskit_experiments/library/stark/ramsey_xy.py:
```diff
@@ -0,0 +1,319 @@
+        """
+        self._timing = None
+
+        super().__init__(qubits=[qubit], analysis=RamseyXYAnalysis(), backend=backend)
```
In the result db the entry show up with experiment name, so user can distinguish the result. I also noticed it is shown as null value. This is weird.

nkanazawa1989 commented:
Thanks @wshanks for your review. After reading your review comment, I added two major changes.

- Added dedicated tutorial for Stark experiment.
- Reorganized module structure. The dedicated module `qiskit_experiments.library.stark` is now absorbed by `qiskit_experiments.library.characterization`. `StarkRamseyXY` is moved to `ramsey_xy`.

coruscating reviewed (commented):
Thanks for writing an experiment manual. Is it possible to show an example run of the experiment and the output figure?

nkanazawa1989 commented:
Thanks @coruscating for reviewing the doc. I don't plan to add any experiment code with this PR. Probably we can write some experiment manual with another follow-up PRs #1009 #1010 .  Note that Stark experiments require dynamics simulation with at least three-level system Hamiltonian. I don't want to introduce another useless mock backend just for writing a tutorial.

coruscating commented:
@nkanazawa1989 One possible solution is to mock some demo code without running it using the `jupyter-input` directive, then show a figure of what the expected output might look like on a real device, just like the example experiment you linked to in the PR description. I think the manual might be a bit unclear right now for new users since it mentions Stark experiments in Qiskit Experiments but doesn't show the name of the experiment class and an example run. It's fine with me either way whether you want to add that in this PR or in one of the follow-up ones.

nkanazawa1989 commented:
Thanks, maybe jupyter-input directive is one of the solutions. If CI can access entry in the result db with public access level, we could load experiment data form there in a hidden cell (I prefer this solution). Anyways this experiment is not really interesting. I started with this experiment because this is very simple while it provides sufficient code example for a reviewer to understand the code structure of Ramsey experiments. The experiment result you will get with this experiment looks identical to `RamseyXY` experiment.

wshanks commented:
Even if the figure just looks like RamseyXY, I think a plot is better than no plot. When trying to use a new experiment, I find it helpful to have a reference for what the data should look like. For some of the qiskit-experiments experiments that have no manual, I come back to the repo and look up their PRs to find a reference. Also, @coruscating didn't mention it but she might partly have been asking because any image would be an improvement here :slightly_smiling_face: : 

![image](https://user-images.githubusercontent.com/1178780/233473515-1c3c7eb8-0569-499e-92f7-8c2b06bfe010.png)

wshanks commented on qiskit_experiments/library/characterization/ramsey_xy.py:
```diff
@@ -208,3 +208,278 @@ def _metadata(self):
+
+
+class StarkRamseyXY(BaseExperiment):
+    """A sign-sensitive experiment to measure the frequency of a qubit under pulsed Stark tone.
```
```suggestion
    """A sign-sensitive experiment to measure the frequency of a qubit under a pulsed Stark tone.
```

wshanks commented on qiskit_experiments/library/characterization/ramsey_xy.py:
```diff
@@ -208,3 +208,278 @@ def _metadata(self):
+                                                                               0
+
+        In principle, the sequence is a variant of :class:`.RamseyXY` circuit.
+        However, the delay in between √X gates is replaced with the off-resonant drive.
```
```suggestion
        However, the delay in between √X gates is replaced with an off-resonant drive.
```

wshanks commented on qiskit_experiments/library/characterization/ramsey_xy.py:
```diff
@@ -208,3 +208,278 @@ def _metadata(self):
+class StarkRamseyXY(BaseExperiment):
+    """A sign-sensitive experiment to measure the frequency of a qubit under pulsed Stark tone.
+
+    # section: overview
```
Should the description link to the manual somewhere (I know it is linked in the options)?

nkanazawa1989 replied:
Ah thanks! I forgot to add the link.

wshanks commented on qiskit_experiments/library/characterization/ramsey_xy.py:
```diff
@@ -208,3 +208,278 @@ def _metadata(self):
+        Args:
+            physical_qubits: Index of physical qubit.
+            backend: Optional, the backend to run the experiment on.
+            experiment_options: Extra experiment options. See ``self.experiment_options``.
```
I am not sure that referencing `self.experiment_options` helps that much. Maybe something like this?

```suggestion
            experiment_options: Experiment options. See the class documentation or ``self._default_experiment_options`` for descriptions.
```

wshanks reviewed (commented):
This is looking good. Sorry for the long wait!

nkanazawa1989 commented on docs/manuals/characterization/stark_experiment.rst:
```diff
@@ -0,0 +1,146 @@
+
+.. math::
+
+    \delta f_S \propto \frac{\alpha}{2\Delta\left(\alpha - \Delta\right)} \Omega^2,
```
This comes from the appendix of effective CR paper https://arxiv.org/pdf/1804.04073.pdf. Empirically the shift is quadratic function of drive, but the fit model also uses cubit term to cover some nonlinear case. Probably `\sim` is better?

![image](https://user-images.githubusercontent.com/39517270/233801502-d42992f6-d75f-4dc6-bd27-28e79cdd3dee.png)

wshanks replied:
Okay, hmm, using `sim` (or `\approx`?) seems okay then. In the tutorial, we are mostly focused on the single qubit case and the corrections to the expression of proportional to `J`, so you could take `J=0` if you wanted. In your first reference they only include the first term:

![image](https://user-images.githubusercontent.com/1178780/233813664-9536a5d0-6969-467e-815f-b4331810254a.png)

nkanazawa1989 replied:
Okey, it's replaced with `\approx`

nkanazawa1989 commented on docs/manuals/characterization/stark_experiment.rst:
```diff
@@ -0,0 +1,146 @@
+
+where :math:`\alpha` is the qubit anharmonicity and :math:`\Delta=f_S - f_0` is the
+frequency separation of the Stark tone from the qubit frequency :math:`f_0`.
+We sometime call this *Stark shift*.
```
I didn't notice this page. Seems like a good reference. This is added as ref3.

nkanazawa1989 commented:
Thanks Will for the second review. In [c3724ba](https://github.com/Qiskit/qiskit-experiments/pull/1008/commits/c3724ba4802ed7681eba0edbed51a4a5d36a1e00) I added new section to explain the pulse technique I employed, which shows a nice thumb nail and gives better understanding of the technique for users who are not familiar with it.

wshanks commented on qiskit_experiments/library/characterization/ramsey_xy.py:
```diff
@@ -208,3 +208,330 @@ def _metadata(self):
+        if stark_freq_offset is not None and stark_sigma is not None:
+            if stark_freq_offset < 1 / stark_sigma:
+                warnings.warn(
+                    "Provided configuration may induce coherence state exchange of qubit "
```
```suggestion
                    "Provided configuration may induce coherent state exchange between qubit levels"
```

wshanks commented on docs/manuals/characterization/stark_experiment.rst:
```diff
@@ -0,0 +1,225 @@
+Following pulse sequence illustrates how :math:`\delta f_S` is characterized
+by a variant of the Hahn-echo pulse sequence [5]_.
+
+.. jupyter-execute::
```
Hmmm, this is better than it saying "NO THUMB NAIL" but I wonder why it is picking up the qiskit logo and not the image generated by this code block? @coruscating any idea?

![image](https://user-images.githubusercontent.com/1178780/234954633-3dc0caab-123a-407f-b8a4-eceb35694263.png)

nkanazawa1989 replied:
![image](https://user-images.githubusercontent.com/39517270/235761503-c066de1c-f886-46f9-b24b-7ae84be910ab.png)

Updated. I needed to specify the figure name in the configuration file.
https://nbsphinx.readthedocs.io/en/0.7.1/gallery/thumbnail-from-conf-py.html

wshanks commented on docs/manuals/characterization/stark_experiment.rst:
```diff
@@ -0,0 +1,225 @@
+This approach may not work for other device architectures.
+
+
+Characterizing Frequency Shift
```
For consistency with the other section titles:

```suggestion
Characterizing the frequency shift
```

wshanks commented on docs/manuals/characterization/stark_experiment.rst:
```diff
@@ -0,0 +1,225 @@
+------------------------------
+
+One can experimentally measure :math:`\delta f_S` with the :class:`.StarkRamseyXY` experiment.
+Following pulse sequence illustrates how :math:`\delta f_S` is characterized
```
```suggestion
The following pulse sequence illustrates how :math:`\delta f_S` is characterized
```

wshanks commented on docs/manuals/characterization/stark_experiment.rst:
```diff
@@ -0,0 +1,225 @@
+As you may notice, this sequence is interleaved with two pulses labeled
+"StarkV" (Gaussian) and "StarkU" (GaussianSquare) filled in yellow, representing Stark tones.
+These pulses are designed to have the same maximum amplitude :math:`\Omega` resulting
+in the same :math:`\delta f_S` at this amplitude -- but why we need two pulses?
```
```suggestion
in the same :math:`\delta f_S` at this amplitude -- but why do we need two pulses?
```

wshanks commented on docs/manuals/characterization/stark_experiment.rst:
```diff
@@ -0,0 +1,225 @@
+These pulses are designed to have the same maximum amplitude :math:`\Omega` resulting
+in the same :math:`\delta f_S` at this amplitude -- but why we need two pulses?
+
+Since :math:`\delta f_S` is amplitude dependent, Stark tones cause time-dependent
+frequency shift at pulse ramps. With a single Stark tone, you are only able to estimate
```
```suggestion
Since :math:`\delta f_S` is amplitude dependent, the Stark pulses cause time-dependent
frequency shifts during the pulse ramps. With a single Stark tone, you are only able to estimate
```

wshanks commented on docs/manuals/characterization/stark_experiment.rst:
```diff
@@ -0,0 +1,225 @@
+the average :math:`\delta f_S` over the history of amplitudes :math:`\Omega(t)`,
+even though you may want to characterize :math:`\delta f_S` at a particular :math:`\Omega`.
+You have to remember that you cannot use a square envelope to set a uniform amplitude,
+because such pulse is known to have broad frequency component.
```
```suggestion
because the sharp rise and fall of the pulse amplitude has a broad frequency spectrum
which could produce unwanted excitations.
```

wshanks commented on docs/manuals/characterization/stark_experiment.rst:
```diff
@@ -0,0 +1,225 @@
+You have to remember that you cannot use a square envelope to set a uniform amplitude,
+because such pulse is known to have broad frequency component.
+
+The pulse sequence shown in above is adopted to address such issue.
```
```suggestion
The pulse sequence shown above is adopted to address such issue.
```

wshanks commented on docs/manuals/characterization/stark_experiment.rst:
```diff
@@ -0,0 +1,225 @@
+The pulse sequence shown in above is adopted to address such issue.
+The Z rotation accumulated by the first pulse is proportional to :math:`\int \Omega_V^2(t) dt`,
+while that of the second pulse is :math:`-\int \Omega_U^2(t) dt` because
+the qubit state to prove this rotation is flipped by the pi-pulse in the middle.
```
Trying to clarify this a little more:

```suggestion
the qubit state is flipped by the pi-pulse in the middle, flipping the sense of rotation of the state even though the actual rotation direction is the same for both pulses.
```

wshanks commented on docs/manuals/characterization/stark_experiment.rst:
```diff
@@ -0,0 +1,225 @@
+The Z rotation accumulated by the first pulse is proportional to :math:`\int \Omega_V^2(t) dt`,
+while that of the second pulse is :math:`-\int \Omega_U^2(t) dt` because
+the qubit state to prove this rotation is flipped by the pi-pulse in the middle.
+The only difference of the :math:`\Omega_U(t)` from the :math:`\Omega_V(t)` is the flat-top part
+with a constant amplitude :math:`\Omega`, where :math:`\delta f_S` is also constant.
```
```suggestion
The only difference between :math:`\Omega_U(t)` and :math:`\Omega_V(t)` is the flat-top part
with constant amplitude :math:`\Omega`, where :math:`\delta f_S` is also constant.
```

wshanks commented on docs/manuals/characterization/stark_experiment.rst:
```diff
@@ -0,0 +1,225 @@
+the qubit state to prove this rotation is flipped by the pi-pulse in the middle.
+The only difference of the :math:`\Omega_U(t)` from the :math:`\Omega_V(t)` is the flat-top part
+with a constant amplitude :math:`\Omega`, where :math:`\delta f_S` is also constant.
+Thanks to this sign flip, the net Z rotation accumulated through two tones is
+only proportional to the flat-top part of the StarkU pulse.
```
```suggestion
Thanks to this sign flip, the net Z rotation accumulated through the two pulses is
proportional to only the flat-top part of the StarkU pulse.
```

wshanks commented on qiskit_experiments/framework/backend_timing.py:
```diff
@@ -180,21 +180,46 @@ def circuits(self):
+        self,
+        backend: Backend,
+        *,
+        acquire_alignment: Optional[int] = None,
```
Hmm, do you think it is useful to have all these arguments? I was thinking just `round_pulse` would have a `enforce_minimum=False` option, so you could create an instance from the backend but get pulse durations not rounded up to the minimum if needed. I don't oppose the argument overrides, but I think the `enforce_minimum` (or a better name) option is still a good idea so that you don't need two `BackendTiming` instances to get both a real pulse duration and an incremental pulse duration.

nkanazawa1989 replied:
This could help us writing a test with some mock backend which doesn't provide constraints. I took an approach to modify constraint value to keep the method signature clean. Regarding ignoring the minimum condition, we need to modify not only `round_pulse` but also `pulse_time`. As we may introduce further controllability in future, e.g. ignore acquire alignment, then method would become much busier. As you mentioned, we need to define multiple instances for different controllability.

wshanks replied:
Hmm, okay. I see your point. We can see how it goes. It might be worth more complexity in the BackendTiming class if it meant less complexity in the code using it.

wshanks commented:
This is good to me now @nkanazawa1989. I pulled it out of the merge queue for now because I think you should update the PR description since that becomes the commit message.

nkanazawa1989 commented:
Thanks Will for catching this. I updated the PR description as well.
